velox/connectors/hive/tests/HivePartitionFunctionTest.cpp

```diff
@@ -246,14 +246,12 @@ TEST_F(HivePartitionFunctionTest, timestamp) {
        Timestamp(
            std::numeric_limits<int64_t>::min(),
            std::numeric_limits<uint64_t>::min()),
-       Timestamp(
-           std::numeric_limits<int64_t>::max(),
-           std::numeric_limits<uint64_t>::max())});
+       Timestamp(std::numeric_limits<int64_t>::max(), Timestamp::kMaxNanos)});
```

Comment:
Is this a valid timestamp? What would it return from Timestamp::toMillis() method?

Reply:
This timestamp.toMillis() will overflow. In fact, these toXXX methods all need some restrictions on seconds. toNanos needs the time represented by timestamp to be less than about 292 years after epoch (INT64_MAX/10^9/60/60/24/365 = 292.4712086775). 
In addition, too small seconds may be meaningless. The smallest meaningful timestamp should be 0001-01-01 00:00:00. But from the perspective of the Timestamp type itself, seconds can take any value, but it cannot be expressed as nanos or may not be able to express a meaningful time.

Reply:
> > toNanos needs the time represented by timestamp to be less than about 292 years after epoch
> 
> Should we define Timestamp::kMaxSeconds and Timestamp::kMinSeconds and use these to check 'seconds' in the constructor?

I think triggering overflow is also acceptable. Although very large seconds cannot be converted to nanos, it is still reasonable for the timestamp type. In Presto, we can create a timestamp larger than 2262-04-12 (1970-01-01 + 292).
```
presto> select from_unixtime(99999999999);
          _col0
-------------------------
 5138-11-16 17:46:39.000
(1 row)
```
And we can also create a negative timestamp, although it is meaningless in actual use.  
```
presto> select from_unixtime(-99999999999);
          _col0
--------------------------
 -1199-02-15 22:19:04.000
(1 row)
```
Timestamp(std::numeric_limits<int64_t>::max(), Timestamp::kMaxNanos) and Timestamp(std::numeric_limits<int64_t>::lowest(), std::numeric_limits<uint64_t>::lowest()) are also useful in min/max agg functions.

In Spark, although they stated in the comments that the range of timestamp is [0001-01-01T00:00:00, 9999-12-31T23:59:59], but we can still create some meaningless timestamps.
```
spark-sql> select from_unixtime(-99999999999);
-1199-02-15 22:19:04
Time taken: 0.064 seconds, Fetched 1 row(s)
spark-sql> select from_unixtime(99999999999);
5138-11-16 17:46:39
Time taken: 0.041 seconds, Fetched 1 row(s)
```

Reply:
I see. Should we document this in the class comments?

Reply:
> I see. Should we document this in the class comments?

Sure.

Reply:
Hi @mbasmanova thanks for mention this PR!

After discuss with @liujiayi771 offline, we found this PR does not fix #5199, because second_ has not limit, and std::gmtime will throw when pass int64_t max value.

Do we fix toString problem?

Reply:
@mbasmanova The root cause of this problem is that Timestamp::toString() uses std::gmtime method. Although std::time_t is typedef int64_t in 64-bit machine, but time_t's tm_year is a 32-bit int, whose maximum value is 292277094. 292277094 * 365 * 24 * 60 * 60 = 9,217,250,436,384,000 $\approx$ int64_max / 1000. So if we enter int64_t value large than 9,217,250,436,384,000, gmtime will return nullptr.

Reply:
In Presto from_unixtime function, if input seconds is too large, it will always return the upper bound.
```
presto:default> select from_unixtime(9223372036854775807);
             _col0
-------------------------------
 +292278994-08-17 15:12:55.807
(1 row)

Query 20230608_070526_00097_hc74z, FINISHED, 1 node
Splits: 17 total, 17 done (100.00%)
22ms [0 rows, 0B] [0 rows/s, 0B/s]

presto:default> select from_unixtime(9223372036854775);
             _col0
-------------------------------
 +292278994-08-17 15:12:55.807
(1 row)
```

Reply:
@liujiayi771 I feel we should not allow creating Timestamps with too large 'seconds'. Would it make sense to cap seconds at some reasonable value?

Reply:
> @liujiayi771 I feel we should not allow creating Timestamps with too large 'seconds'. Would it make sense to cap seconds at some reasonable value?

I test Presto's from_unixtime, the range of seconds is [-INT64_MAX/1000, +INT64_MAX/1000].  According to the implementation of Presto, it stores the timestamp in milliseconds, uses double to do calculation, and finally truncates when the return is converted to long.
```
public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime)
{
    return Math.round(unixTime * 1000);
}
``` 

But Velox's timestamp is not stored in milliseconds, so there is no problem with a larger range than Presto. Most of the timestamp ranges of different systems are related to the data type they used. The boundary of this range does not have much meaning for actual use, it is just the definition of the system itself. Do you think Velox is also limited to [-INT64_MAX/1000, +INT64_MAX/1000]?

Reply:
I think it is useful to make sure that if Timestamp is created successfully, then all methods of the Timestamp class can be called without any error. Hence, it seems useful to put a limit on seconds.

Reply:
Hmm... I wonder if toNanos method itself makes sense. Timestamp type uses 2 64-bit integers, but this method suggests that the value can be packed into a single int64_t. If that was true, then why would one use 2 64-bit integers and waste memory? Should we remove this method altogether?

```
  int64_t toNanos() const {
    // int64 can store around 292 years in nanos ~ till 2262-04-12
    // The addition cannot overflow because the product will be promoted to
    // uint64_t first and its value is at most UINT64_MAX / 2.
    return checkedMultiply(seconds_, (int64_t)1'000'000'000) + nanos_;
  }
```

Reply:
The core reason why Velox uses two int64s to represent timestamp is to make the precision reach nanosecond level. In fact int96 is enough, nanos can use uint32 to represent. 

toNanos is a basic method of timestamp if its precision is nanoseconds. In actual data, timestamp is not too large. There are still application scenarios for converting to nanoseconds. I think it should not be removed. Throwing an exception is a better choice.
For example:
DuckDB throw exception: https://github.com/duckdb/duckdb/blob/bdb75246b9aaee5f971e8511e8e635309ec231fb/src/common/types/timestamp.cpp#L339

Impala return false for overflow:
https://github.com/apache/impala/blob/80c1d2dbaabc78bdf1b6e4da5475bfa365cd375e/be/src/runtime/timestamp-value.inline.h#L175

clickhouse throw exception:
https://github.com/ClickHouse/ClickHouse/blob/d0f031660bac854190843ac431f0665b8034967c/src/Functions/FunctionUnixTimestamp64.h#L73

DuckDB's Timestamp::toString also throws exception when out of bounds
https://github.com/duckdb/duckdb/blob/bdb75246b9aaee5f971e8511e8e635309ec231fb/src/common/types/timestamp.cpp#L285

I still prefer not to limit the seconds, or limit according to whether it can be converted to milliseconds. Large seconds is not wrong for timestamp, but it just cannot be converted into nanoseconds and cannot be converted into a formatted string.

My point is:
1. limit seconds to [-INT64_MAX/1000, +INT64_MAX/1000]. This can at least ensure that the precision of timestamp is not lower than Presto, and can make Timestamp::toString work.
2. not limit seconds, and toString throw exception when it cannot create the corret format string.

toNanos just throw exception if overflow.

Reply:
@liujiayi771 Thank you for detailed explanation. Makes sense. How about we do (1) above?

> limit seconds to [-INT64_MAX/1000, +INT64_MAX/1000]. This ensures that the range of timestamp will not be smaller than Presto, and can make Timestamp::toString work.